lib/optimizer/src/Transformer/PreloadHeroImage.php

```diff
@@ -0,0 +1,563 @@
+            }
+
+            return new HeroImage(
+                $src,
+                $childNode->getAttribute(Attribute::MEDIA),
+                $childNode->getAttribute(Attribute::SRCSET),
+                $childNode
+            );
+        }
+
+        return null;
+    }
+
+    /**
+     * Generate the preload link for a given hero image.
+     *
+     * @param HeroImage       $heroImage     Hero image to generate the preload link for.
+     * @param Document        $document      Document to generate the preload link in.
+     * @param ErrorCollection $errors        Collection of errors that are collected during transformation.
+     * @param DOMNode|null    $referenceNode Reference node after which to insert the preload link. Null if none.
+     */
+    private function generatePreload(
+        HeroImage $heroImage,
+        Document $document,
+        ErrorCollection $errors,
+        DOMNode $referenceNode = null
+    ) {
+        if ($heroImage->getSrcset()) {
+            $errors->add(Error\CannotPreloadImage::fromImageWithSrcsetAttribute($heroImage->getAmpImg()));
+            return;
+        }
+
+        if ($this->hasExistingImagePreload($document, $heroImage->getSrc())) {
+            return;
+        }
+
+        $preload = $document->createElement(Tag::LINK);
+        $preload->setAttribute(Attribute::REL, Attribute::REL_PRELOAD);
+        $preload->setAttribute(Attribute::HREF, $heroImage->getSrc());
+        $preload->setAttribute('as', 'image');
+        $preload->setAttribute(Attribute::DATA_HERO, null);
+
+        if ($heroImage->getMedia()) {
+            $preload->setAttribute(Attribute::MEDIA, $heroImage->getMedia());
+        }
+
+        if ($referenceNode) {
+            $referenceNode->parentNode->insertBefore($preload, $referenceNode->nextSibling);
+        } else {
+            $document->head->appendChild($preload);
+        }
+    }
+
+    /**
+     * Generate the SSR image element for the hero image.
+     *
+     * @param HeroImage $heroImage Hero image to generate the SSR image element for.
+     * @param Document  $document  Document in which to generate the SSR image element in.
+     */
+    private function generateImg(HeroImage $heroImage, Document $document)
+    {
+        $element = $heroImage->getAmpImg();
+
+        if (! $element) {
+            return;
+        }
+
+        $imgElement = $document->createElement(Tag::IMG);
+        $imgElement->setAttribute(Attribute::CLASS_, self::SSR_IMAGE_CLASS);
+        $imgElement->setAttribute(Attribute::DECODING, 'async');
+
+        foreach (self::ATTRIBUTES_TO_COPY as $attribute) {
+            if ($element->hasAttribute($attribute)) {
+                $imgElement->setAttribute($attribute, $element->getAttribute($attribute));
+            }
+        }
+
+        $element->setAttribute(Attribute::I_AMPHTML_SSR, null);
+        $element->setAttribute(Attribute::DATA_HERO, null);
+
+        $element->appendChild($imgElement);
```

Comment:
```suggestion
        $element->appendChild($imgElement);

        // Remove any noscript>img when an amp-img is pre-rendered.
        $noscript = $document->xpath->query('./noscript[ img ]', $element)->item(0);
        if ($noscript instanceof DOMElement) {
            $noscript->parentNode->removeChild($noscript);
        }
```

See https://github.com/ampproject/amphtml/blob/44ea165c34f9cce37979f2849a88ca90016d568a/validator/validator-main.protoascii#L1709-L1751